spec/index.html

```diff
@@ -1493,3 +1503,221 @@ <h1>SharedArray ( ..._values_ )</h1>
+      <li>is not intended to be called as a function and will throw an exception when called in that manner.</li>
+    </ul>
+
+    <emu-clause id="sec-atomics.mutex">
+      <h1>Atomics.Mutex ( )</h1>
+      <p>This function performs the following steps when called:</p>
+      <emu-alg>
+        1. If NewTarget is *undefined*, throw a *TypeError* exception.
+        1. Let _createInitializer_ be a new Abstract Closure with parameters (_newMutex_) that captures nothing and performs the following steps when called:
+          1. Set _newMutex_.[[MutexWaiterList]] to a new WaiterList Record.
+          1. Set _newMutex_.[[IsLocked]] to *false*.
+          1. Return ~unused~.
+        1. Return SharedStructCreate(_createInitializer_, « [[MutexWaiterList]], [[IsLocked]] »).
+      </emu-alg>
+    </emu-clause>
+  </emu-clause>
+
+  <emu-clause id="sec-properties-of-the-mutex-constructor">
+    <h1>Properties of the Mutex Constructor</h1>
+    <p>The Mutex constructor:</p>
+    <ul>
+      <li>has a [[Prototype]] internal slot whose value is %Function.prototype%.</li>
+      <li>has the following properties:</li>
+    </ul>
+
+    <emu-note>
+      <p>Per-Realm prototypes, which is currently an open design question and not included in this draft, will give Mutexes a per-Realm prototype with built-in methods instead of static methods, and will also add support for the %Symbol.dispose%.</p>
+    </emu-note>
+
+    <emu-clause id="sec-atomics.mutex.lock">
+      <h1>Atomics.Mutex.lock ( _mutex_ [ , _unlockToken_ [ , _timeout_ ] ] )</h1>
+      <p>This function puts the surrounding agent in a wait queue and suspends it until the mutex is unlocked or until the wait times out, returning a String differentiating those cases.</p>
+      <p>It performs the following steps when called:</p>
+      <emu-alg>
+        1. If _mutex_ is not an Object, throw a *TypeError* exception.
+        1. If _mutex_ does not have a [[MutexWaiterList]] internal slot, throw a *TypeError* exception.
+        1. If _unlockToken_ not *undefined*, then
+          1. If _unlockToken_ is not an Object, throw a *TypeError* exception.
+          1. If _unlockToken_ does not have a [[LockedMutex]] internal slot, throw a *TypeError* exception.
+          1. If _unlockToken_.[[LockedMutex]] is not ~empty~, throw a *TypeError* exception.
+        1. If _timeout_ is *undefined*, set _timeout_ to *+∞*<sub>𝔽</sub>.
+        1. If _timeout_ is not a Number, throw a *TypeError* exception.
+        1. If _timeout_ is either *NaN* or *+∞*<sub>𝔽</sub>, let _t_ be +∞; else if _timeout_ is *-∞*<sub>𝔽</sub>, let _t_ be 0; else let _t_ be max(ℝ(_timeout_), 0).
+        1. If AgentCanSuspend() is *false*, throw a *TypeError* exception.
+        1. Let _result_ be LockMutex(_mutex_, _t_).
+        1. If _result_ is ~acquired~, then
+          1. Return UnlockTokenCreateIfNeeded(_unlockToken_, _mutex_).
+        1. Else,
+          1. Assert: _result_ is ~timed-out~.
+          1. Return *"timed-out"*.
```

Comment:
I expect we'll eventually make tokens disposable, which makes the common pattern look like

```js
using token = mutex.lock(timeout);
```
Returning a string here means that will throw, which is... probably the thing you want? It'll be kind of a cryptic error message, but that's better than thinking you have the lock when you don't (`using token = null` is legal, so returning `null` here would make that silently succeed).

Reply:
I'm against throwing because:

1. A timeout isn't an exceptional thing though, but one of the expected uses of the proposal.
2. The method can throw for parameter validation reasons. Are you asking users to check instanceof? Do string comparison on the error message?

That said the string thing is not the cleanest, but I don't know how else to express it. The constraints are:

- We need a distinguishing timeout value
- We can't allocate a result object on every call

Open to ideas.

Reply:
In conjunction with the `wait` and `waitFor` split below, I've made `Mutex.lock` return `UndefinedToken` when the lock is acquired, or `null` when timed out. WDYT?

Reply:
Yeah, `null` has this problem. The options I see are:

- stop using tokens
- return `null` or the token and have this footgun
- make tokens not disposable
- allocate an additional object so you can do `{ success: boolean, token?: UnlockToken }`
- have some static frozen placeholder object, like `Mutex.TIMED_OUT = Object.freeze({})`, and return that
- return a string and just live with the mixed return type

Of these, while I share @rbuckton's distaste for the mixed return type, I think the last is the best option.

Reply:
> Of these, while I share @rbuckton's distaste for the mixed return type, I think the last is the best option.

Agreed, and I reverted the commit that made it `null`.

Extra allocation is definitely out of the question, for the same reason why all these APIs allow token reuse.

I considered a static placeholder object, but that's just more steps for the same effect. Presumably the distaste for mixing types is the return type is more complicated, being `|"timed-out"`. Using a static frozen object still complicates it, since you _really_ don't want to generalize the return type to `Object`, which would help no one, so you'd get `UnlockToken|TimedOutTokenSingleton` or, which is a worse way to say `"timed-out"`.

Reply:
I think it would make more sense to move the timeout to `tryLock`/`lockIfAvailable` since a lock with a timeout is just another case of "attempt to take a lock". The difference is that a `tryLock` without a timeout (or perhaps with a timeout of `0`) indicates that the method should return immediately if the lock couldn't be taken, while any other value indicates a blocking call until the timeout elapses.

Reply:
Let me noodle on it for a little bit. I'm liking the timeout split idea like I did for Condition.

Reply:
It is very easy to look at the code

```js
try {
  using void = mutex.tryLock(timeout);
  somethingWhichNeedsTheLock();
} catch {
  // ...
}
```
and assume that it is correct because `tryLock` will throw if you don't get the lock. And if you happen to get lucky and actually do get the lock during testing, you won't notice. Which means you'll get data corruption instead of an error in prod.

That would be a bit better if it were called `lockIfAvailable` because `try` is associated with throwing and `if` is not, but I still think strings are better because then you'd actually get an error. It should be hard to have code which looks like it gets a lock but does not actually get a lock.

Also, smaller point, `tryLock` needs to be usable on the main thread; `Atomics.wait` currently isn't even with 0 timeout. But we could change `Atomics.wait`, presumably.

Reply:
I have the exact opposite intuition. I would assume `tryLock` would *not* throw if you fail to get the lock. I would expect `lock` with a timeout to throw if you fail to get the lock.

If you had a `lock` that took a timeout and threw if that timeout elapsed, then I would expect `tryLock` to be roughly analogous to:

```js
tryLock(timeout) {
  try {
    this.lock(timeout);
    return true;
  }
  catch (e) {
    if (e instanceof TimeoutError) {
      return false;
    }
    throw e;
  }
}
```

Thus the `try` in `tryLock` here is doing the work of the `try..catch` (and related filtering) on your behalf (and likely more efficiently than the above as well).

A method like `tryLock`, `tryAdd`, `tryCreate`, etc. generally *shouldn't* throw (except maybe for argument validation or truly exceptional cases). That's generally what the word "try" implies, and even underpins what `try..catch` and `try.finally` do.

> That would be a bit better if it were called `lockIfAvailable` because `try` is associated with throwing and `if` is not.

I'd much rather have `set.tryAdd` than `set.addIfNotPresent` purely for the more concise spelling. I can honestly say I would use it often enough that the longer spelling would be a major irritant, even with completions in an editor (and perhaps especially due to completions in an editor, since it would always appear after `add`).

> Also, smaller point, `tryLock` needs to be usable on the main thread; `Atomics.wait` currently isn't even with 0 timeout. But we could change `Atomics.wait`, presumably.

IIRC, we expect that `tryLock` could be used on the main thread even with a timeout, though the runtime could potentially clamp the timeout to a maximum value or perhaps throw for a non-zero timeout as part of argument validation.

Reply:
It's perfectly reasonable for you to have the intuition that `tryLock` would not throw, but I stand by my claim that it is easy for a reader who is not already familiar with what `tryLock` does to look at the buggy code in my comment and not realize that it's bugged, in the worst possible way. The API doesn't just need to be hard for you personally to misuse catastrophically; it should be hard for other people to misuse catastrophically as well.

> `set.addIfNotPresent`

I think it's OK to for the atomics operations to have slightly more verbose naming conventions. The tradeoff between convenience and misuse-resistance is very different for atomics than for stuff like `Set`.

Sidebar: I have no idea what `set.addIfNotPresent` is supposed to do. "Add if not present" is how `set.add` already works.

> `tryLock` could be used on the main thread even with a timeout

That's not consistent with the last point in [the readme of the microwaits proposal](https://github.com/tc39/proposal-atomics-microwait?tab=readme-ov-file#i-still-want-to-block-the-main-thread-can-we-clamp-timeouts-on-atomicswait) and would be a weird inconsistency. I'm happy to allow both it and `Atomics.wait` on the main thread with 0 timeout, or even potentially with a clamped or bounded timeout, but I don't see any reason to allow one but not the other.

Anyway, as long as it works at least with 0 timeout on the main thread, and is consistent with `Atomics.wait`, that seems fine.

Reply:
> Sidebar: I have no idea what `set.addIfNotPresent` is supposed to do. "Add if not present" is how `set.add` already works.

This is moving further off topic so I'll summarize and we can discuss offline if necessary. What I'd want is a `tryAdd(value: T): boolean` that does the work of `.has()` and `.add()` in one step. I've always found it unfortunate that `Set.prototype.add()` returns `this` rather than something informative such as whether the value was added (`true`) or was already present (`false`), given that my most frequent uses for `Set` involve this pattern.

Reply:
> tryLock could be used on the main thread even with a timeout

Only if the timeout is 0, non-0 clamped timeouts are too hard to do IMO for the same reasons that I abandoned it in `Atomics.pause`.

I'm going with the string-returning version for the spec draft. Please open an issue for the alternative and let's hammer it out during Stage 2.

Reply:
Chatted through with Ron and understand better where he's coming from: that resource-returning methods being conditional was always part of the explicit design of `using`. I'd like to hear committee's thoughts on the footgun here. Given Ron's a co-champion, I'll put `return null` for the timeout case and call it out as an open question.

Reply:
I see the footgun, let's discuss it in plenary.